model/point_model.py

```diff
@@ -113,6 +119,62 @@ def set_pulse_shape(
         """
         self._pulse_generator = pulse_generator
 
+    def add_noise(
+        self,
+        noise_to_signal_ratio: float,
+        seed: Union[None, int] = None,
+        noise_type: str = "additive",
+    ) -> None:
+        """
+        Specifies noise for realization.
+        Parameters
+        ----------
+        noise_to_signal_ratio: float, defined as X_rms/S_rms where X is noise and S is signal.
+        seed: None or int, seed for the noise generator
+        noise_type: str
+            "additive": additive noise
+            "dynamic": dynamic noise (only applicable for constant duration times)
+            "both": both additive and dynamic noise
+        """
+        assert noise_type in {"additive", "dynamic", "both"}
+        assert seed is None or isinstance(seed, int)
+        assert noise_to_signal_ratio >= 0
+
+        self._noise_type = noise_type
+        self._noise_random_number_generator = np.random.RandomState(seed=seed)
+        mean_amplitude = self._forcing_generator.get_forcing(
+            self._times, gamma=self.gamma
+        ).amplitudes.mean()
+        self._sigma = np.sqrt(noise_to_signal_ratio * self.gamma) * mean_amplitude
+
+        self._noise = np.zeros(len(self._times))
+
+    def _discretize_noise(self) -> np.ndarray:
+        """Discretizes noise for the realization"""
+
+        if self._noise_type in {"additive", "both"}:
+            self._noise += self._sigma * self._noise_random_number_generator.normal(
+                size=len(self._times)
+            )
+
+        if self._noise_type in {"dynamic", "both"}:
+            durations = self._last_used_forcing.durations
```

Comment:
Yes, I understand it works fine now, but in this way, someone could easily break it unintentionally by just changing the order of the lines. I think we should implement it in a way that is not so fragile, right?

Reply:
Sorry I don't think I fully understand your point. 
`_discretize_noise()` gets called in `make_realization()` so it is always the same (correct) realization of the forcing as the one used for `_add_pulse_to_signal()`.

> someone could easily break it unintentionally by just changing the order of the lines

Could you please specify which lines you mean?

Reply:
self._last_used_forcing = forcing and _discretize_noise()

Reply:
`_discretize_noise()` is meant to be a protected method and shouldn't be called by the user directly.

My idea for the API was the following:
```Python
model = pm.PointModel(...)
model.add_noise(...)
times, signal = model.make_realization()
```
Such an example is included in `model/example.py` (line 27-29).

What do you think of this design?

Reply:
Yeah, I agree, that's the intended design, and the user of the API will use it correctly. I am worried about the developer, not the user.

Reply:
the forcing is now given to `_discretize_noise(forcing)` as an input parameter instead of using `self._last_used_forcing`. Is that less error prone now in your opinion?

Reply:
Yes, Ill approve